test/e2e/storage/testsuites/multivolume.go

```diff
@@ -503,3 +527,62 @@ func TestConcurrentAccessToSingleVolume(f *framework.Framework, cs clientset.Int
 		utils.CheckReadFromPath(f, pod, *pvc.Spec.VolumeMode, path, byteLen, seed)
 	}
 }
+
+// getCurrentTopologies() goes through all Nodes and returns unique driver topologies and count of Nodes per topology
+func getCurrentTopologiesNumber(cs clientset.Interface, keys []string) ([]topology, []int, error) {
```

Comment:
Instead of returning a slice for topologies and another for the count, I think it'd be better to simply return a `map[string]int` of `topologyKey` -> count.

At line 577 we're iterating over `topologyList` and using its index in `topologyCount`, with a `map` we don't have to do that.

Reply:
I think this would be trickier: `topology` is a non-comparable data type (we even need a special `topologyEqual` function) so I can't create a map from it.

Reply:
I see, I didn't realize `topology` is a `map[string]string` already.

If I may suggest, we could consider using a different data structure instead. This would allow us to replace these 2 functions by something like:

```
// ensureTopologyRequirements sets nodeSelection affinity according to given topology keys for drivers that provide them.
// If no node with the given topology key is found, this function will skip the test directly.
func ensureTopologyRequirements(nodeSelection *e2epod.NodeSelection, cs clientset.Interface, topologyKeys []string, minCount int) error {
	nodes, err := e2enode.GetReadySchedulableNodes(cs)
	if err != nil {
		return err
	}

	// Get the number of nodes per zone per topology key.
	// Example: map["topology.ebs.csi.aws.com/zone"]["us-east-1"] = 2
	topologyCount := map[string]map[string]int{}
	for _, node := range nodes.Items {
		for _, topologyKey := range topologyKeys {
			if zone, ok := node.Labels[topologyKey]; ok {
				if _, ok := topologyCount[topologyKey]; !ok {
					topologyCount[topologyKey] = make(map[string]int)
				}
				topologyCount[topologyKey][zone]++
			}
		}
	}

	// Once we find a topology with at least minCount of nodes, we use it in the tests
	for topologyKey, zoneCount := range topologyCount {
		for zone, count := range zoneCount {
			if count >= minCount {
				e2epod.SetNodeAffinityTopologyRequirement(nodeSelection, topology{topologyKey: zone})
				return nil
			}
		}
	}

	framework.Skipf("No topology with at least %d nodes found - skipping", minCount)
	return nil
}

```